synapse/util/tracerutils.py

```diff
@@ -0,0 +1,309 @@
+    SPAN_KIND_PRODUCER = INVALID_TAG
+    SPAN_KIND_RPC_CLIENT = INVALID_TAG
+    SPAN_KIND_RPC_SERVER = INVALID_TAG
+
+
+def only_if_tracing(func):
+    """Executes the function only if we're tracing. Otherwise return.
+    Assumes the function wrapped may return None"""
+
+    @wraps(func)
+    def f(*args, **kwargs):
+        if _opentracing:
+            return func(*args, **kwargs)
+        else:
+            return
+
+    return f
+
+
+_opentracing = None
+
+# Block everything by default
+_homeserver_whitelist = None
+
+tags = _DumTagNames
+
+
+def init_tracer(config):
+    """Set the whitelists and initialise the JaegerClient tracer
+
+    Args:
+        config (Config)
+        The config used by the homserver. Here it's used to set the service
+        name to the homeserver's.
+    """
+    if not config.tracer_config.get("tracer_enabled", False):
+        # We don't have a tracer
+        return
+
+    import_opentracing()
+    setup_tags()
+    setup_tracing(config)
+
+
+def import_opentracing():
+    global _opentracing
+    try:
+        # Try to import the tracer. If it's not there we want to throw an eror
+        import opentracing
+    except ImportError as e:
+        logger.error(
+            "The server has been configure to use opentracing but "
+            "the {} module has not been installed.".format(e.name)
+        )
+        raise
+
+    _opentracing = opentracing
+
+
+def setup_tracing(config):
+    try:
+        from jaeger_client import Config as JaegerConfig
+        from synapse.util.scopecontextmanager import LogContextScopeManager
+    except ImportError as e:
+        logger.error(
+            "The server has been configure to use opentracing but "
+            "the {} module has not been installed.".format(e.name)
+        )
+        raise
+
+    # Include the worker name
+    name = config.worker_name if config.worker_name else "master"
+
+    set_homeserver_whitelist(config.tracer_config["homeserver_whitelist"])
+    jaeger_config = JaegerConfig(
+        config={"sampler": {"type": "const", "param": 1}, "logging": True},
+        service_name="{} {}".format(config.server_name, name),
+        scope_manager=LogContextScopeManager(config),
+    )
+    jaeger_config.initialize_tracer()
+
+
+@only_if_tracing
+def setup_tags():
+    global tags, _opentracing
+    tags = _opentracing.tags
+
+
+# Could use kwargs but I want these to be explicit
+@only_if_tracing
+def start_active_span(
+    operation_name,
+    child_of=None,
+    references=None,
+    tags=None,
+    start_time=None,
+    ignore_active_span=False,
+    finish_on_close=True,
+):
+    # We need to enter the scope here for the logcontext to become active
+    _opentracing.tracer.start_active_span(
+        operation_name,
+        child_of=child_of,
+        references=references,
+        tags=tags,
+        start_time=start_time,
+        ignore_active_span=ignore_active_span,
+        finish_on_close=finish_on_close,
+    ).__enter__()
```

Comment:
Can we just return this if opentracing is enabled, and return a noop context manager if it doesn't? A noop context manager would look like:

```python
@contextlib.contextmanager
def _noop_context_manager(*args, **kwargs):
    yield
```